### DISCUSSION TITLE
watch VS rerenders again..

### DISCUSSION THREAD

**Discussion by npup** [react-hook-form/react-hook-form#9752] [Q&A] [answered]
I recently received excellent help from @leapful in another Q&A/discussion, and I hope many others can benefit from the suggestion and explanation he gave there: https://github.com/react-hook-form/react-hook-form/discussions/9638
This post is about a related problem.

The other discussion was about wrapping a form field component (MUI datepicker in this case) to split its value into two parts in the main form. I needed help in triggering the validation from within the wrapped component's onchange as well as regular submit, and got a nice [CSB](https://codesandbox.io/s/bold-currying-olpv3w?file=/src/components/DateTimePicker.jsx) that showed a working solution for this.

The general (and working) solution for wrapping and splitting a form field value as described above involved creating an inner context, watch its changes and use `setValue` on the outer context.  However, this also awakens the watch/rerender beast.
There are some solutions for problems in this area posted elsewhere, but I cannot seem to apply them in this case.  The rerender occurrs, and does not seem to be memoized-away, for example.  Here is a CSB with a commented-out watch ready to wreck everything :) -> https://codesandbox.io/s/bold-babycat-cdnxfs?file=/src/views/Example.jsx:601-602

Any suggestion on how to make the values dynamically observable is greatly appreciated.

PS. the "Tip" widget seemed broken in the previous discussion when I wanted to use it.  Will try again though :)

**Comment by leapful** [comment, marked as answer]
@npup 

I think any data that can be created newly each re-rendering like your `FormContext` (variable `formCtx` with `setValidator`) and `DateTimePicker` component (prop `rules`) should be memorized too.

Working demo: https://codesandbox.io/s/kind-voice-nwnhgu?file=/src/views/Example.jsx

**Reply by npup** [reply]
Thank you again for your explanation and even a working example, that helped a lot.
Even though my example CSB was a simplified example of my real use case 😅 , I didn't quite detect the mutation hazards of those properties with regard to rerendering.  Thank you ever so much for pointing them out.
